lightning/src/offers/offer.rs

```diff
@@ -102,13 +120,21 @@ impl OfferBuilder {
 	/// while the offer is valid.
 	///
 	/// Use a different pubkey per offer to avoid correlating offers.
-	pub fn new(description: String, signing_pubkey: PublicKey) -> Self {
+	pub fn new(description: String, signing_pubkey: SigningPubkey) -> Self {
```

Comment:
When using the `Explicit` variant, the metadata will only be generated if `metadata_derived` is called. Two use cases:
- LDK-independent usage (must call `metadata`)
- LDK usage with public node id (must call `metadata_derived`)

When using `Derived`, the metadata will be generated without needing to call anything. See tests.

Reply:
Note that, as written, if a user gives `SigningPubkey::Explicit` and calls `OfferBuilder::metadata_derived` with an LDK-independent `ExpandedKey`, then `InvoiceRequest::verify` should work correctly. So it could be useful leaving this exposed for general usage.

Reply:
You'd have to ensure that the nonce used to create the signing pubkey is the same one passed to `OfferBuilder::metadata_derived`. With the current approach the builder ensures that it is the same.

Reply:
We'd have to make `OfferContents::metadata` an `Option` if we did so.

Reply:
> What if we instead move the `signing_pubkey` argument and the metadata type argument to the final `build` method? Then we could make it all one argument - either explicit for both or implicit for both.

Note that `metadata` is optional, so it isn't required when given an explicit `signing_pubkey`. Also, that approach doesn't support verification for when an explicit `signing_pubkey` is given (node id or otherwise), FWIW.

> Can we use a `BuildingOffer` struct or so?

Meh, not sure if I really want to repeat the fields and need to do something similar in all the other types, which may have nested contents. Seems error-prone.

> The current API just seems really confusing - users have to provide an enum variant at `OfferBuilder` constructor which can be explicit, in which case they can call the manual metadata function, or it can be derived, in which case they cant (or dont need to) call either metadata function, and if they provide derived they _cant_ call the derived-metadata function (unless its with the same data) or they'll get the wrong pubkey at the end?

Hmmm... if we think it will be confusing, then I'm more inclined to keep the constructor as it was and add a `pub(crate)` constructor for the derived variant taking `&ExpandedKey` and `Nonce`.

Then the question would be whether or not we want to support verification externally. If no, then `OfferBuilder::metadata_derived` is only needed for when used internally with a node id. So we can either make it `pub(crate)` or replace it with another `pub(crate)` constructor.

My feeling is that LDK-independent users likely would just want to set the `signing_pubkey` and `metadata` directly and won't care about the derived versions or verification. If true, I'd lean towards the `pub(crate)` approached outlined above.